apps/designer/app/shared/router-utils/user-content-params.ts

```diff
@@ -0,0 +1,90 @@
+export type CanvasRouteMode = "edit" | "preview" | "published";
+const modes = ["edit", "preview", "published"] as CanvasRouteMode[];
+const getMode = (url: URL): CanvasRouteMode => {
+  const modeParam = url.searchParams.get("mode");
+  const mode =
+    modeParam === null ? "published" : modes.find((mode) => mode === modeParam);
+  if (mode === undefined) {
+    throw new Error(`Invalid mode "${modeParam}"`);
+  }
+  return mode;
+};
+
+// A subtype of Request. To make testing easier.
+type MinimalRequest = {
+  url: string;
+  headers: { get: (name: string) => string | null };
+};
+
+const getRequestHost = (request: MinimalRequest): string =>
+  request.headers.get("x-forwarded-host") || request.headers.get("host") || "";
+
+export const getUserContentBaseUrl = (request: MinimalRequest): string => {
```

Comment:
seems like it would be nicer to pass the env as an argument, given how the tests look like

Reply:
Feels weird to pass env since it's globally available. Maybe just to simplify tests?

Reply:
I get it, though its in the end better then all the mocking strategies I have seen so far, it can be optional and default to process.env, so normal consumers don't have to pass it